velox/functions/prestosql/ArrayAllMatch.cpp

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/functions/lib/LambdaFunctionUtil.h"
+#include "velox/functions/lib/RowsTranslationUtil.h"
+#include "velox/vector/FunctionVector.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+class AllMatchFunction : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VELOX_CHECK_EQ(args.size(), 2);
+    VELOX_CHECK_EQ(TypeKind::BOOLEAN, outputType->kind());
+    VELOX_CHECK_EQ(TypeKind::ARRAY, args[0]->typeKind());
+    VELOX_CHECK_EQ(TypeKind::FUNCTION, args[1]->typeKind());
+
+    exec::LocalDecodedVector arrayDecoder(context, *args[0], rows);
+    auto& decodedArray = *arrayDecoder.get();
+
+    auto flatArray = flattenArray(rows, args[0], decodedArray);
+    auto inputOffsets = flatArray->rawOffsets();
+    auto inputSizes = flatArray->rawSizes();
+
+    std::vector<VectorPtr> lambdaArgs = {flatArray->elements()};
+    auto newNumElements = flatArray->elements()->size();
+
+    SelectivityVector finalSelection;
+    if (!context.isFinalSelection()) {
+      finalSelection = toElementRows<ArrayVector>(
+          newNumElements, *context.finalSelection(), flatArray.get());
+    }
+
+    VectorPtr matchBits;
+    auto elementToTopLevelRows = getElementToTopLevelRows(
+        newNumElements, rows, flatArray.get(), context.pool());
+
+    // Loop over lambda functions and apply these to elements of the base array,
+    // in most cases there will be only one function and the loop will run once.
+    context.ensureWritable(rows, BOOLEAN(), result);
+    auto flatResult = result->asFlatVector<bool>();
+    exec::LocalDecodedVector bitsDecoder(context);
+    auto it = args[1]->asUnchecked<FunctionVector>()->iterator(&rows);
+    while (auto entry = it.next()) {
+      auto elementRows = toElementRows<ArrayVector>(
+          newNumElements, *entry.rows, flatArray.get());
+      auto wrapCapture = toWrapCapture<ArrayVector>(
+          newNumElements, entry.callable, *entry.rows, flatArray);
+
+      entry.callable->apply(
```

Comment:
We may need to add some logic to properly handle errors. For example, if there is an array [1, 2, 0, 3] and predicate x -> (10 / x) < 5, then all_match should return false and "suppress" division by zero error.

```
presto> select all_match(array[1, 2, 0, 3], x -> (10 / x) > 5);
 _col0
-------
 false
(1 row)
```

Reply:
IMHO, it seems complicated to handle and suppress the user exception here. Because it applies the lambda function to multiple rows, we need to handle/suppress the user exception (eg, the `VELOX_ARITHMETIC_ERROR` ) and set the result vector correctly in terms of the type of the results row by row in the expression evaluation framework, which would be complicated and out of the scope of this pr. 
Can we not do it for now or leave a TODO here?

Reply:
It does complicate the logic, but it is necessary to provide correct implementation. Check out the implementation of TRY (EvalCtx::throwOnError) and ConjunctExpr::evalSpecialForm. It might be possible to use the same mechanism here. Also see https://github.com/facebookincubator/velox/issues/3993#issuecomment-1450239804

Reply:
I've added error handling based on your guidance and the [error handling part of expression](https://facebookincubator.github.io/velox/develop/expression-evaluation.html#error-handling-in-and-or-try:~:text=AND%2C%20OR%2C%20TRY-,%C2%B6,-When%20evaluating%20AND) in the doc. It turns out that the expression framework already has a facility for handling errors at a row granularity.